### PR TITLE
Fix sample reports retrieval

### PR DESCRIPTION
**Pull request by ramonski**
## Description of the issue/feature this PR addresses

**Please merge https://github.com/senaite/senaite.core/pull/2279 first**

This PR introduces the following new getters for samples:

- `getRawReports`: Lookup sample reports from the report catalog 
- `getReports`: Returns sample report objects
- `getReportUIDs`: Returns a list of sample report UIDs

## Current behavior before PR

Sample reports were assumed to live inside the sample and were retrieved via `objectValues`

## Desired behavior after PR is merged

Sample reports may live in other samples and *contain* only the current sample in a multi report. Therefore, a catalog search is required.

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/bika/lims/content/analysisrequest.py [hunk 1]
```diff
@@ -1645,26 +1646,61 @@ def getLate(self):
                 return True
         return False
 
+    def getRawReports(self, inlcude_provisional=True, **kw):
```
Comment (xispa) [inline src/bika/lims/content/analysisrequest.py]: We usually use the `getRaw*` function to return UIDs. I suggest to add the `full_objects=False` in `getReports` (along with **kwargs) instead.

Also note that `getRawReports` could rely on `get_backreferences`:

```python
def getRawReports(self):
    return get_backreferences(self, "ARReportAnalysisRequest")

def getReportUIDs(self):
    return self.getRawReports()

```

Also note there is a typo with `inlcude_provisional` (should be `include_provisional`)

```python
def getReports(self, include_provisional=True, full_objects=False, **kw):
    ...
```
```

File: src/bika/lims/content/analysisrequest.py [hunk 2]
```diff
@@ -1645,26 +1646,61 @@ def getLate(self):
...
-        report_list = sorted(self.objectValues('ARReport'),
-                             key=lambda report: report.getDatePublished())
+
+        report_list = self.getRawReports()
         if not report_list:
             return "0"
+
         last_report = report_list[-1]
-        if last_report.getDatePrinted():
+        if last_report.getDatePrinted:
```
Comment (xispa) [inline src/bika/lims/content/analysisrequest.py]: OMG, this "Printed" fake-workflow needs to be revisited at some point... and removed probably

**Review by xispa — CHANGES_REQUESTED** [review]
I suggest renaming `getRawReports` to be compliant with AT behavior that we adopted for all types (DX included)

File: src/senaite/core/upgrade/v02_05_000.py [hunk 3]
```diff
@@ -129,3 +133,22 @@ def setup_catalogs(tool):
...
+    """
+    logger.info("Update report catalog ...")
+    portal = api.get_portal()
+
+    # ensure new indexes are created
+    setup_catalog_mappings(portal)
+    setup_core_catalogs(portal)
+
+    # remove columns
+    del_column(REPORT_CATALOG, "getClientTitlegetClientURL")
```
Comment (xispa) [inline src/senaite/core/upgrade/v02_05_000.py:149]: Oops, should be:

```python
del_column(REPORT_CATALOG, "getClientTitle")
del_column(REPORT_CATALOG, "getClientURL")
```

Comment (ramonski) [reply src/senaite/core/upgrade/v02_05_000.py:149]: No, it is actually correct. Note the missing `,` in the catalog.

File: src/bika/lims/content/analysisrequest.py (hunk 2 above)
Comment (ramonski) [reply src/bika/lims/content/analysisrequest.py]: Yes indeed. This is kind of ... well, I would not say useless, but ....

File: src/senaite/core/upgrade/v02_05_000.py (hunk 3 above)
Comment (xispa) [reply src/senaite/core/upgrade/v02_05_000.py:149]: Oops, you are right

File: src/bika/lims/content/analysisrequest.py (hunk 1 above)
Comment (ramonski) [reply src/bika/lims/content/analysisrequest.py]: Yes, I thought about this. Shall I rely better on the backreferences instead?

Comment (xispa) [reply src/bika/lims/content/analysisrequest.py]: > Yes, I thought about this. Shall I rely better on the backreferences instead?

I think that relying on backreferences is better, yes

**Review by xispa — APPROVED** [review]

**Pull request merged**